presto-main/src/main/java/com/facebook/presto/cost/CachingCostProvider.java

```diff
@@ -65,48 +78,221 @@ public CachingCostProvider(CostCalculator costCalculator, StatsProvider statsPro
 
-        PlanNodeCostEstimate cumulativeCost = localCosts.add(sourcesCost);
-        return cumulativeCost;
+        MemoryEstimate memoryEstimate = node.accept(new PeakMemoryEstimator(), new EstimationContext(localCost.getMemoryCost(), sourcesCosts));
+        verify(memoryEstimate != null);
+
+        PlanCostEstimate sourcesCost = sourcesCosts.stream()
+                .reduce(PlanCostEstimate.zero(), CachingCostProvider::addSiblingsCost);
+
+        return new PlanCostEstimate(
+                localCost.getCpuCost() + sourcesCost.getCpuCost(),
+                Math.max(sourcesCost.getMaxMemory(), Math.max(memoryEstimate.maxMemoryBeforeOutputting, memoryEstimate.maxMemoryWhenOutputting)),
+                memoryEstimate.maxMemoryWhenOutputting,
+                localCost.getNetworkCost() + sourcesCost.getNetworkCost());
+    }
+
+    private static PlanCostEstimate addSiblingsCost(PlanCostEstimate a, PlanCostEstimate b)
+    {
+        return new PlanCostEstimate(
+                a.getCpuCost() + b.getCpuCost(),
+                a.getMaxMemory() + b.getMaxMemory(),
+                a.getMaxMemoryWhenOutputting() + b.getMaxMemoryWhenOutputting(),
+                a.getNetworkCost() + b.getNetworkCost());
+    }
+
+    private static class EstimationContext
+    {
+        private final double localMaxMemory;
+        private final List<PlanCostEstimate> sourcesCosts;
+
+        public EstimationContext(double localMaxMemory, List<PlanCostEstimate> sourcesCosts)
+        {
+            this.localMaxMemory = localMaxMemory;
+            this.sourcesCosts = requireNonNull(sourcesCosts, "sourcesCosts is null");
+        }
+
+        @Override
+        public String toString()
+        {
+            return toStringHelper(this)
+                    .add("localMaxMemory", localMaxMemory)
+                    .add("sourcesCosts", sourcesCosts)
+                    .toString();
+        }
+    }
+
+    private static class MemoryEstimate
+    {
+        private final double maxMemoryBeforeOutputting;
+        private final double maxMemoryWhenOutputting;
+
+        public MemoryEstimate(double maxMemoryBeforeOutputting, double maxMemoryWhenOutputting)
+        {
+            this.maxMemoryBeforeOutputting = maxMemoryBeforeOutputting;
+            this.maxMemoryWhenOutputting = maxMemoryWhenOutputting;
+        }
+
+        @Override
+        public String toString()
+        {
+            return toStringHelper(this)
+                    .add("maxMemoryBeforeOutputting", maxMemoryBeforeOutputting)
+                    .add("maxMemoryWhenOutputting", maxMemoryWhenOutputting)
+                    .toString();
+        }
+    }
+
+    private static class PeakMemoryEstimator
+            extends PlanVisitor<MemoryEstimate, EstimationContext>
+    {
+        @Override
+        protected MemoryEstimate visitPlan(PlanNode node, EstimationContext context)
+        {
+            if (context.sourcesCosts.isEmpty()) {
```

Comment:
Is this just tablescans and values nodes?  I'd rather have explicit methods for those.

Reply:
Maybe i was just lazy.... OTOH even `com.facebook.presto.cost.CostCalculatorUsingExchanges.CostEstimator#visitPlan` has "generic" handling (or lack thereof) of many node types.
I will try to get rid of it.

Reply:
> I will try to get rid of it.

Just note: getting rid of this shortcut may lead to having many copy-paste-like methods.

Reply:
I'm ok leaving it if leaf types would always have this kind of cost, just leave a comment explaining why.